server/etcdserver/server.go

```diff
@@ -892,6 +903,26 @@ func (s *EtcdServer) revokeExpiredLeases(leases []*lease.Lease) {
 	})
 }
 
+// ensureLeadership checks whether current member is still the leader.
+func (s *EtcdServer) ensureLeadership() bool {
+	lg := s.Logger()
+	if err := s.linearizableReadNotify(s.ctx); err != nil {
```

Comment:
`linearizableReadNotify` does two things:
* Get the readIndex
* Wait for readIndex to be applied to db

For this purpose we only need the first one and having the second part might be risky as it injects an arbitrary log wait. I see two solutions:
* Limit the wait by adding timeout to context.
* Separate readIndex notification like in https://github.com/etcd-io/etcd/pull/16792

Reply:
added a timeout.  

We will backpprt the PR to 3.5 and 3.4, so let's try to keep the change as simple as possible.